File: cpp/121_Best_Time_to_Buy_and_Sell_Stock.cpp
#include <vector>

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int length = prices.size();
        if (length <= 1)
            return 0;
        vector<int> buys(length);   //第i天或者第i天前买入的最少花费
        vector<int> sells(length);  //第i天或者第i天后卖出的最大收益
        buys[0] = prices[0];
        sells[length-1] = prices[length-1];
        for (int i=1; i<length; i++)
            buys[i] = min(buys[i-1], prices[i]);
        for (int i=length-2; i >= 0; i--)
            sells[i] = max(sells[i+1], prices[i]);
        int max_profit = 0;
        for (int i=0; i<length; i++)
            max_profit = max(max_profit, sells[i] - buys[i]);
        return max_profit;
    }
};

File: python/121.Best Time to Buy and Sell Stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) == 0:
            return 0
        buys = [0] * len(prices)
        sells = [0] * len(prices)
        buys[0] = prices[0]
        sells[len(prices)-1] = prices[len(prices)-1]
        for i in range(1, len(prices)):
            buys[i] = min(prices[i], buys[i-1])
        for i in range(len(prices)-2, -1, -1):
            sells[i] = max(prices[i], sells[i+1])
        ret = 0
        for i in range(len(prices)):
            if (sells[i] - buys[i] > ret):
                ret = sells[i] - buys[i]
        return ret